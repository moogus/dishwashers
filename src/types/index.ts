export type GridItem = {
  productId: string;
  price: string;
  title: string;
  image: string;
};

export type AdditionalServices = {
  includedServices: string[];
}

export type Product = {
  title: string;
  media: Images,
  price: string;
  productInformation: string;
  displaySpecialOffer: string;
  additionalServices: AdditionalServices;
  code: string;
}

export type Price = {
  now?: string
}

export type ApiGridItem = {
  productId: string;
  price: Price;
  title: string;
  image: string;
}

export type Images = {
  altText: string;
  urls: string[]
}

export type Media = {
  images: Images;
}

export type Details = {
  productInformation: string;
}

export type ApiProduct = {
  title : string;
  media: Media;
  price?: Price;
  details?: Details;
  displaySpecialOffer : string;
  additionalServices: AdditionalServices;
  code : string;
}