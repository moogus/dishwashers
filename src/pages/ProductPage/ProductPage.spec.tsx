import React from "react";
import { render } from "@testing-library/react";
import { mocked } from "ts-jest/utils";

import useLoadProduct from "../../hooks/product/load-product";
import useGetProduct from "../../hooks/product/get-product";

import { ProductModel } from "../../types";

import { ProductPage } from "./index";

jest.mock("../../hooks/product/load-product");
jest.mock("../../hooks/product/get-product");
jest.mock("@reach/router");

describe("<ProductPage />", () => {
  const renderContainer = ({ productId = "foo" } = {}) => {
    const props = { productId, path: "foo-path (needed for reach router)" };
    return render(<ProductPage {...props} />);
  };

  describe("when product api is loading", () => {
    beforeEach(() => {
      mocked(useLoadProduct).mockClear();
      mocked(useGetProduct).mockClear();
      mocked(useLoadProduct).mockReturnValue({
        loading: true,
        error: false,
      });
      mocked(useGetProduct).mockReturnValue({} as ProductModel);
    });

    test("a loading message is displayed", () => {
      const { container } = renderContainer();
      expect(container).toMatchSnapshot();
    });
  });

  describe("when product api has loaded", () => {
    beforeEach(() => {
      mocked(useLoadProduct).mockClear();
      mocked(useGetProduct).mockClear();
      mocked(useLoadProduct).mockReturnValue({
        loading: false,
        error: false,
      });
      mocked(useGetProduct).mockReturnValue(({
        title: "foo",
        media: {
          altText: "what the dickens",
          urls: ["foo"],
        },
        price: "Â£1000",
        details: { productInformation: "<p>hiya</p>" },
        displaySpecialOffer: "baz",
        additionalServices: { includedServices: ["stuff"] },
        code: "12345",
      } as unknown) as ProductModel);
    });

    test("a product page is rendered as expected", () => {
      const { container } = renderContainer();
      expect(container).toMatchSnapshot();
    });
  });
});
