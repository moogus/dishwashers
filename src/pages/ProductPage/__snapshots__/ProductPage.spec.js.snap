// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProductPage /> when product api has loaded a product page is rendered as expected 1`] = `
<div>
  <div
    class="sc-idOhPF XgPF"
  >
    <header>
      <div
        class="d-flex justify-content-start"
      >
        <button
          class="sc-dIUggk MVjvb"
        >
          ‹
        </button>
        <h1
          class="sc-bdfBwQ eChoRV pt-4 pb-4"
        >
          foo
        </h1>
        <h1
          class="sc-fFubgz jgOXvX pt-4 pb-4"
        >
          title
        </h1>
      </div>
    </header>
    <main>
      <section>
        <div
          class="d-none d-sm-block d-md-block d-lg-none"
        >
          <div
            class="sc-jrAGrp kOzzJp row"
          >
            <div
              class="sc-iqHYGH ebGHqy col"
            >
              <div
                class="row"
              >
                <div
                  class="sc-dlfnbm bARfyf col"
                >
                  <img
                    alt="what the dickens"
                    class="sc-gsTCUz kPBfKS d-flex justify-content-center"
                    data-testid="selected-image"
                    src="foo"
                  />
                </div>
              </div>
              <div
                class="row"
              >
                <div
                  class="col"
                >
                  <div
                    class="d-flex justify-content-center"
                    data-testid="image-select-buttons"
                  >
                    <button
                      class="sc-hKgILt eHwdyN"
                    >
                      •
                    </button>
                  </div>
                </div>
              </div>
              <div
                class="sc-kEjbxe gFqndP row pt1 pb1 border-top border-bottom"
              />
              <div
                class="sc-gKsewC gCFaFZ row"
              >
                <div
                  class="col"
                >
                  <p
                    class="sc-jSgupP iMfPPR"
                  >
                    £1000
                  </p>
                  <p
                    class="sc-iBPRYJ hrHJix"
                  >
                    Claim an extra 3 years guarantee via redemption
                  </p>
                  <p>
                    stuff
                  </p>
                </div>
              </div>
              <div
                class="sc-crrsfI clIMvm row"
              >
                <div
                  class="col"
                >
                  <h2
                    class="sc-fubCfw bXzuaz"
                  >
                    Product Information
                  </h2>
                </div>
              </div>
              <div
                class="sc-crrsfI clIMvm row"
              >
                <div
                  class="col"
                >
                  <h2
                    class="sc-pFZIQ jwQzaS"
                  >
                    Product code: 12345
                  </h2>
                </div>
              </div>
              <div
                class="sc-crrsfI clIMvm row"
              >
                <div
                  class="col"
                >
                  <span />
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="d-none d-lg-block -xl-block"
        >
          <div
            class="sc-kstrdz mTIWh"
          >
            <div
              class="sc-hBEYos bctvzG"
            >
              <div
                class="row"
              >
                <div
                  class="col-8  pr-0"
                >
                  <div
                    class="row"
                  >
                    <div
                      class="sc-dlfnbm bARfyf col"
                    >
                      <img
                        alt="what the dickens"
                        class="sc-gsTCUz kPBfKS d-flex justify-content-center"
                        data-testid="selected-image"
                        src="foo"
                      />
                    </div>
                  </div>
                  <div
                    class="row"
                  >
                    <div
                      class="col"
                    >
                      <div
                        class="d-flex justify-content-center"
                        data-testid="image-select-buttons"
                      >
                        <button
                          class="sc-hKgILt eHwdyN"
                        >
                          •
                        </button>
                      </div>
                    </div>
                  </div>
                  <div
                    class="sc-fodVxV ihlQgI row"
                  >
                    <div
                      class="col"
                    >
                      <h2
                        class="sc-dQppl bKQiMl"
                      >
                        Product Information
                      </h2>
                    </div>
                  </div>
                  <div
                    class="sc-fodVxV ihlQgI row"
                  >
                    <div
                      class="col "
                    >
                      <h2
                        class="sc-bqyKva giqUCo"
                      >
                        Product code: 12345
                      </h2>
                    </div>
                  </div>
                  <div
                    class="sc-fodVxV ihlQgI row"
                  >
                    <div
                      class="col"
                    >
                      <span />
                    </div>
                  </div>
                </div>
                <div
                  class="col-4 border-left"
                >
                  <div
                    class="sc-gKsewC gCFaFZ row"
                  >
                    <div
                      class="col"
                    >
                      <p
                        class="sc-jSgupP iMfPPR"
                      >
                        £1000
                      </p>
                      <p
                        class="sc-iBPRYJ hrHJix"
                      >
                        Claim an extra 3 years guarantee via redemption
                      </p>
                      <p>
                        stuff
                      </p>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </section>
    </main>
  </div>
</div>
`;

exports[`<ProductPage /> when product api is loading a loading message is displayed 1`] = `
<div>
  <div
    class="sc-bkzZxe bvuktt"
  >
    Loading Dishwashers
  </div>
</div>
`;
